Python/Product/PythonTools/PythonTools/Debugger/DebugAdapter/DebugAdapterProcess.cs

```diff
@@ -46,36 +47,38 @@ internal sealed class DebugAdapterProcess {
             _targetHostProcess = _targetInterop.ExecuteCommandAsync(pythonExePath, "\"" + _debuggerAdapterDirectory + "\"");
 
             if (!string.IsNullOrEmpty(webBrowserUrl) && Uri.TryCreate(webBrowserUrl, UriKind.RelativeOrAbsolute, out _)) {
-                _timer = new Timer(OnBrowserLaunchElapsedTimer, null, 5000, Timeout.Infinite);
+                _adapterHostContext.Events.PreviewProtocolEvent += MonitorLaunchBrowserMessage;
             }
 
             return _targetHostProcess;
         }
 
-        private void OnBrowserLaunchElapsedTimer(object state) {
-            Uri.TryCreate(_webBrowserUrl, UriKind.RelativeOrAbsolute, out var uri);
-            OnPortOpenedHandler.CreateHandler(uri.Port, null, null, () => _targetHostProcess.HasExited, LaunchBrowserDebugger);
-            _timer.Dispose();
-        }
-
-        private void LaunchBrowserDebugger() {
-            var vsDebugger = (IVsDebugger2)VisualStudio.Shell.ServiceProvider.GlobalProvider.GetService(typeof(SVsShellDebugger));
+        private void MonitorLaunchBrowserMessage(object sender, PreviewProtocolEventEventArgs e) {
+            if (e.Event.Type.Equals("output") &&
+                e.Event is OutputEvent message &&
+                Regex.Matches(message.Output, ".*(https?:\\/\\/\\S+:[0-9]+\\/?).*").Count == 1
```

Comment:
you should store a `Regex` as static member of this class, created with option `RegexOptions.Compiled`